File: examples/python/challenges/get_leaderboard.py
"""
Get Challenge Leaderboard

Retrieves and displays the leaderboard for a specific challenge.

Usage:
    python get_leaderboard.py

Requirements:
    pip install requests
"""

import requests
import os
from datetime import datetime


def get_challenge_leaderboard(challenge_token, cookie):
    """
    Fetch and display challenge leaderboard.

    Args:
        challenge_token (str): The challenge token/ID
        cookie (str): Your _ncfa cookie value

    Returns:
        list: List of leaderboard entries
    """
    try:
        url = f'https://www.geoguessr.com/api/v3/results/highscores/{challenge_token}'

        response = requests.get(url, cookies={'_ncfa': cookie})
        response.raise_for_status()

        data = response.json()

        # Extract and format leaderboard
        leaderboard = []
        for index, item in enumerate(data['items']):
            player = item['game']['player']
            leaderboard.append({
                'rank': index + 1,
                'username': player['nick'],
                'score': int(player['totalScore']['amount']),
                'is_pro': player['isProUser'],
                'country': player['countryCode'],
                'played_at': datetime.fromisoformat(
                    item['game']['created'].replace('Z', '+00:00')
                ).strftime('%Y-%m-%d %H:%M:%S')
            })

        # Display results
        print(f"\n🏆 Challenge Leaderboard ({len(data['items'])} players)\n")
        print(f"{'Rank':<6} {'Username':<20} {'Score':<10} {'Pro':<5} {'Country'}")
        print("-" * 60)

        for entry in leaderboard[:20]:  # Show top 20
            pro_indicator = "✓" if entry['is_pro'] else " "
            print(f"{entry['rank']:<6} {entry['username']:<20} "
                  f"{entry['score']:<10,} {pro_indicator:<5} {entry['country']}")

        # Statistics
        total_players = len(leaderboard)
        avg_score = sum(p['score'] for p in leaderboard) / total_players
        top_score = leaderboard[0]['score'] if leaderboard else 0

        print(f"\n📊 Statistics:")
        print(f"Total Players: {total_players}")
        print(f"Average Score: {avg_score:,.0f}")
        print(f"Top Score: {top_score:,}")
        print(f"Winner: {leaderboard[0]['username']}")

        return leaderboard

    except requests.exceptions.HTTPError as e:
        print(f"❌ HTTP Error: {e}")
        return None
    except requests.exceptions.RequestException as e:
        print(f"❌ Request Error: {e}")
        return None


def main():
    # Get credentials from environment variables
    cookie = os.getenv('GEOGUESSR_COOKIE')
    if not cookie:
        print("❌ Error: GEOGUESSR_COOKIE environment variable not set")
        print("Set it with: export GEOGUESSR_COOKIE='your_cookie_value'")
        return

    # Example challenge token (replace with your own)
    challenge_token = input("Enter challenge token (or URL): ").strip()

    # Extract token from URL if full URL provided
    if '/' in challenge_token:
        challenge_token = challenge_token.split('/')[-1]

    # Fetch leaderboard
    leaderboard = get_challenge_leaderboard(challenge_token, cookie)

    if leaderboard:
        print("\n✅ Successfully retrieved leaderboard!")


if __name__ == '__main__':
    main()

File: examples/javascript/challenges/get-leaderboard.js
/**
 * Get Challenge Leaderboard
 *
 * Retrieves and displays the leaderboard for a specific challenge.
 *
 * Usage in browser console:
 * 1. Navigate to geoguessr.com (must be logged in)
 * 2. Paste this code into the console
 * 3. Call: getChallengeLeaderboard('YOUR_CHALLENGE_TOKEN')
 */

async function getChallengeLeaderboard(challengeToken) {
    try {
        const url = `https://www.geoguessr.com/api/v3/results/highscores/${challengeToken}`;

        const response = await fetch(url, {
            credentials: 'include'
        });

        if (!response.ok) {
            throw new Error(`HTTP ${response.status}: ${response.statusText}`);
        }

        const data = await response.json();

        // Extract and format leaderboard
        const leaderboard = data.items.map((item, index) => {
            const player = item.game.player;
            return {
                rank: index + 1,
                username: player.nick,
                score: parseInt(player.totalScore.amount),
                isProUser: player.isProUser,
                countryCode: player.countryCode,
                playedAt: new Date(item.game.created).toLocaleString()
            };
        });

        // Display results
        console.log(`\n🏆 Challenge Leaderboard (${data.items.length} players)\n`);
        console.table(leaderboard);

        // Additional stats
        const totalPlayers = data.items.length;
        const avgScore = leaderboard.reduce((sum, p) => sum + p.score, 0) / totalPlayers;
        const topScore = leaderboard[0]?.score || 0;

        console.log(`\n📊 Statistics:`);
        console.log(`Total Players: ${totalPlayers}`);
        console.log(`Average Score: ${Math.round(avgScore)}`);
        console.log(`Top Score: ${topScore}`);
        console.log(`Winner: ${leaderboard[0]?.username}`);

        return leaderboard;

    } catch (error) {
        console.error('❌ Error fetching leaderboard:', error.message);
        return null;
    }
}

// Example usage:
// getChallengeLeaderboard('6G9h2UPctmUmUtaa');

// Extract token from URL:
// const url = "https://www.geoguessr.com/challenge/6G9h2UPctmUmUtaa";
// const token = url.split('/').pop();
// getChallengeLeaderboard(token);
